modules/library/metadata/src/main/java/org/geotools/factory/FactoryRegistry.java

```diff
@@ -152,15 +158,52 @@ public FactoryRegistry(final Class<?>[] categories) {
      * @param categories The categories.
      */
     public FactoryRegistry(final Collection<Class<?>> categories) {
-        super(categories.iterator());
-        for (final Iterator<Class<?>> it=getCategories(); it.hasNext();) {
-            if (needScanForPlugins == null) {
-                needScanForPlugins = new HashSet<Class<?>>();
-            }
-            needScanForPlugins.add(it.next());
+        for (Class<?> category : categories) {
+            needScanForPlugins.add(category);
+            registry.registerCategory(category);
         }
     }
 
+    /*
+     * TODO: discuss how much API can be broken
+     *
+     * Breakage:
+     *
+     *  - not all methods of `ServiceRegistry` API were implemented
+     *
+     * Possible breakage:
+     *
+     *  - in general nomenclature, replace "ServiceProvider" with "Factory"
+     *  - RegisterableService interface can no longer be used because it expects a `ServiceProvider`
+     *  - replace `Filter` with `Predicate<? super T>`
+     *
+     */
+
+    public Stream<Class<?>> streamCategories() {
+        return registry.streamCategories();
+    }
+
+    // TODO: ensure uniform null handling here and in CategoryRegistry
+
+    // TODO: document
+    public <T> T getServiceProviderByClass(Class<T> providerClass) {
+        if (providerClass == null) {
+            // TODO: do something fancy like in getServiceProvider(Class, Filter, Hints, Hints.Key) ?
```

Comment:
I meant something like [in `registerFromSystemProperty`](https://github.com/nicolaiparlog/geotools/blob/aefb899a7e092f6cd749e58bfa768d449a07d7dc/modules/library/metadata/src/main/java/org/geotools/factory/FactoryRegistry.java#L990):

```
throw new FactoryRegistryException(Errors.format(
ErrorKeys.CANT_CREATE_FACTORY_$1, classname), exception);
```

I thought that may be GeoTools' preferred way of creating error messages and wondered whether I should do something similar.